packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts

```diff
@@ -0,0 +1,607 @@
+  /** Epoch -> Set<blockRoot> */
+  private readonly epochIndex = new MapDef<Epoch, Set<RootHex>>(() => new Set<string>());
+  private readonly metrics: Metrics["cpStateCache"] | null | undefined;
+  private readonly logger: Logger;
+  private readonly clock: IClock | null | undefined;
+  private preComputedCheckpoint: string | null = null;
+  private preComputedCheckpointHits: number | null = null;
+  private readonly maxEpochsInMemory: number;
+  private readonly persistentApis: CPStatePersistentApis;
+  private readonly shufflingCache: ShufflingCache;
+  private readonly getHeadState?: GetHeadStateFn;
+
+  constructor(
+    {metrics, logger, clock, shufflingCache, persistentApis, getHeadState}: PersistentCheckpointStateCacheModules,
+    opts: PersistentCheckpointStateCacheOpts
+  ) {
+    this.cache = new MapTracker(metrics?.cpStateCache);
+    if (metrics) {
+      this.metrics = metrics.cpStateCache;
+      metrics.cpStateCache.size.addCollect(() => {
+        let persistCount = 0;
+        let inMemoryCount = 0;
+        const memoryEpochs = new Set<Epoch>();
+        const persistentEpochs = new Set<Epoch>();
+        for (const [key, cacheItem] of this.cache.entries()) {
+          const {epoch} = fromCheckpointKey(key);
+          if (isPersistedCacheItem(cacheItem)) {
+            persistCount++;
+            persistentEpochs.add(epoch);
+          } else {
+            inMemoryCount++;
+            memoryEpochs.add(epoch);
+          }
+        }
+        metrics.cpStateCache.size.set({type: CacheType.persisted}, persistCount);
+        metrics.cpStateCache.size.set({type: CacheType.inMemory}, inMemoryCount);
+        metrics.cpStateCache.epochSize.set({type: CacheType.persisted}, persistentEpochs.size);
+        metrics.cpStateCache.epochSize.set({type: CacheType.inMemory}, memoryEpochs.size);
+      });
+    }
+    this.logger = logger;
+    this.clock = clock;
+    if (opts.maxEpochsInMemory < 0) {
+      throw new Error("maxEpochsInMemory must be >= 0");
+    }
+    this.maxEpochsInMemory = opts.maxEpochsInMemory;
+    // Specify different persistentApis for testing
+    this.persistentApis = persistentApis;
+    this.shufflingCache = shufflingCache;
+    this.getHeadState = getHeadState;
+  }
+
+  /**
+   * Reload checkpoint state keys from the last run.
+   */
+  async init(): Promise<void> {
+    const persistedKeys = await this.persistentApis.readKeys();
+    for (const persistedKey of persistedKeys) {
+      const cp = this.persistentApis.persistedKeyToCheckpoint(persistedKey);
+      this.cache.set(toCheckpointKey(cp), {type: CacheType.persisted, value: persistedKey});
+    }
+  }
+
+  /**
+   * Get a state from cache, it may reload from disk.
+   * This is an expensive api, should only be called in some important flows:
+   * - Validate a gossip block
+   * - Get block for processing
+   * - Regen head state
+   */
+  async getOrReload(cp: CheckpointHex): Promise<CachedBeaconStateAllForks | null> {
```

Comment:
can this method reuse `getStateOrBytes`?

Reply:
I refactor to a new method `getStateOrLoadDb()` to be used by `getOrReload()` and `getStateOrBytes` please have a look